File: src/helmizer.py
#!/usr/bin/env python3
"""
Helmizer - Generates `kustomization.yaml` for your locally-rendered YAML manifests,
such as from Helm templates or plain YAML manifests
"""
import argparse
import json
import logging
import subprocess
from os import path
from os import walk
from sys import exit
from sys import stdout

import confuse
import yaml
from confuse.exceptions import NotFoundError
from validators import url as validate_url


class Kustomization:
    def __init__(self, helmizer_config, arguments) -> None:
        self.helmizer_config = helmizer_config
        self.arguments = arguments
        self.yaml: dict = dict()

        # get apiVersion
        self.yaml["apiVersion"] = self.get_api_version()

        # get kind
        self.yaml["kind"] = "Kustomization"

        # get strings
        for list_of_keys in ["namePrefix", "nameSuffix", "namespace"]:
            try:
                key_string: str = self.get_str(list_of_keys)
                if key_string:
                    self.yaml[list_of_keys] = key_string
            except NotFoundError:
                pass

        # get lists
        for list_of_keys in [
            "buildMetadata",
            "configMapGenerator",
            "images",
            "patches",
            "patchesJson6902",
            "replacements",
            "replicas",
            "secretGenerator",
            "vars",
        ]:
            try:
                key_list = self.get_list(list_of_keys)
                if key_list:
                    self.yaml[list_of_keys] = key_list
            except NotFoundError:
                pass

        # get dicts
        for list_of_keys in ["commonAnnotations", "commonLabels", "generatorOptions", "openapi", "sortOptions"]:
            try:
                key_dict = self.get_dict(list_of_keys)
                if key_dict:
                    self.yaml[list_of_keys] = key_dict
            except NotFoundError:
                pass

        # get files
        for list_of_keys in ["crds", "components", "patchesStrategicMerge", "resources"]:
            try:
                key_paths = self.get_files(arguments, list_of_keys)
                if key_paths:
                    self.yaml[list_of_keys] = key_paths
            except NotFoundError:
                pass

    def sort_keys(self, arguments) -> bool:
        if arguments.no_sort_keys:
            return True
        else:
            try:
                sort_keys: bool = self.helmizer_config["helmizer"]["sort-keys"].get(bool)
                if sort_keys:
                    for key in ["crds", "components", "patchesStrategicMerge", "resources"]:
                        try:
                            self.yaml[key].sort()
                        except KeyError:
                            pass
                    logging.debug("keys sorted")
            except NotFoundError:
                for key in ["crds", "components", "patchesStrategicMerge", "resources"]:
                    try:
                        self.yaml[key].sort()
                    except KeyError:
                        pass
            finally:
                return True

    def print_kustomization(self) -> bool:
        try:
            print(yaml.dump(self.yaml, sort_keys=False))
        except KeyError:
            print(yaml.dump(self.yaml, sort_keys=False))
        finally:
            return True

    def write_kustomization(self, arguments) -> bool:
        try:
            if arguments.dry_run:
                logging.debug("Performing dry-run, not writing to a file system")
                return True
            elif self.helmizer_config["helmizer"]["dry-run"].get(bool):
                logging.debug("Performing dry-run, not writing to a file system")
                return True
        except NotFoundError:
            pass

        # identify kustomization file's parent directory
        kustomization_directory: str = path.dirname(path.abspath(path.normpath(arguments.helmizer_config)))

        # identify kustomization file name
        kustomization_file_name: str = str()
        try:
            kustomization_file_name = self.helmizer_config["helmizer"]["kustomization-file-name"].get(str)
        except NotFoundError:
            kustomization_file_name = "kustomization.yaml"

        # write to file
        try:
            kustomization_file_path: str = path.normpath(f"{kustomization_directory}/{kustomization_file_name}")
            with open(kustomization_file_path, "w") as file:
                file.write(yaml.dump(self.yaml))
                logging.debug(f"Successfully wrote to file: {path.abspath(kustomization_file_path)}")
            return True
        except IsADirectoryError as e:
            raise e
        except TypeError as e:
            raise e

    def render_template(self, arguments) -> bool:
        logging.debug("Rendering template")
        self.sort_keys(arguments)
        self.print_kustomization()
        self.write_kustomization(arguments)
        return True

    def get_api_version(self) -> str:
        api_version: str = str()
        try:
            api_version = self.helmizer_config["kustomize"]["apiVersion"].get(str)
        except NotFoundError:
            api_version = "kustomize.config.k8s.io/v1beta1"
        finally:
            logging.debug(f"apiVersion: {api_version}")
            return api_version

    def get_files(self, arguments, key: str) -> list:
        target_paths: list = list()
        final_target_paths: list = list()
        kustomization_path: str = str()

        try:
            # test if the key to configure is even defined in input helmizer config
            kustomization_children: list = self.helmizer_config["kustomize"][key].get(list)
            kustomization_directory: str = str()
            try:
                kustomization_directory = self.helmizer_config["helmizer"]["kustomization-directory"].get(str)
            except NotFoundError:
                kustomization_directory = "."

            kustomization_path = path.dirname(
                path.abspath(
                    path.normpath(
                        path.join(
                            arguments.helmizer_config,
                            kustomization_directory,
                        )
                    )
                )
            )

            if len(kustomization_children) > 0:
                # each target path
                for target_path in kustomization_children:
                    child_path: str = path.abspath(path.join(kustomization_path, target_path))

                    # walk directory
                    if path.isdir(child_path):
                        for dirpath, _, filenames in walk(child_path):
                            for filename in filenames:
                                target_paths.append(path.join(dirpath, filename))

                    # file
                    elif path.isfile(child_path):
                        target_paths.append(child_path)

                    # url
                    elif validate_url(child_path):
                        target_paths.append(child_path)

                # remove any ignored files
                try:
                    # walk directory to remove multiple files
                    for ignore in self.helmizer_config["helmizer"]["ignore"].get(list):
                        ignore_abspath: str = path.abspath(path.join(kustomization_path, ignore))
                        if path.isdir(ignore_abspath):
                            for dirpath, _, filenames in walk(ignore_abspath):
                                for filename in filenames:
                                    file_path: str = path.join(dirpath, filename)
                                    logging.debug(f"Removing ignored file from final list: {file_path}")
                                    target_paths.remove(file_path)
                        # remove a file
                        else:
                            logging.debug(f"Removing ignored file from final list: {path.join(kustomization_path, ignore)}")
                            target_paths.remove(path.join(kustomization_path, ignore))  # just one file
                except ValueError:
                    pass
                except NotFoundError:
                    pass

                # convert absolute paths into paths relative to the kustomization directory
                for final_target_path in target_paths:
                    final_target_paths.append(path.relpath(final_target_path, kustomization_path))

                return final_target_paths

        except NotFoundError:
            logging.debug(f"key not found: {key}")
        except KeyError:
            logging.debug(f"key not found: {key}")
        except TypeError:
            logging.debug(f"key not of valid type: {key}")
        finally:
            return final_target_paths

    def get_dict(self, key: str) -> dict:
        raw_yaml: dict = dict()
        try:
            if self.helmizer_config["kustomize"][key].get(dict):
                raw_yaml = json.loads(json.dumps(self.helmizer_config["kustomize"][key].get(dict)))
                logging.debug(f"{key}: {raw_yaml}")
        except NotFoundError:
            logging.debug(f"key not found: {key}")
            return raw_yaml
        except KeyError:
            logging.debug(f"key not found: {key}")
            return raw_yaml
        except TypeError:
            logging.debug(f"key not of valid type: {key}")
            return raw_yaml
        return raw_yaml

    def get_str(self, key: str) -> str:
        raw_yaml: str = str()
        try:
            if self.helmizer_config["kustomize"][key].get(str):
                raw_yaml = json.loads(json.dumps(self.helmizer_config["kustomize"][key].get(str)))
                logging.debug(f"{key}: {raw_yaml}")
        except NotFoundError:
            logging.debug(f"key not found: {key}")
        except KeyError:
            logging.debug(f"key not found: {key}")
        except TypeError:
            pass
        return raw_yaml

    def get_list(self, key: str) -> list:
        raw_yaml: list = list()
        try:
            if len(self.helmizer_config["kustomize"][key].get(list)) > 0:
                raw_yaml = json.loads(json.dumps(self.helmizer_config["kustomize"][key].get(list)))
                logging.debug(f"{key}: {raw_yaml}")
        except NotFoundError:
            logging.debug(f"key not found: {key}")
        except KeyError:
            logging.debug(f"key not found: {key}")
        except TypeError:
            logging.debug(f"key not of valid type: {key}")
        return raw_yaml


def run_subprocess(helmizer_config, arguments):
    # TODO return type checking
    subprocess_working_directory = path.dirname(path.abspath(path.normpath(arguments.helmizer_config)))
    logging.debug(f"Subprocess working directory: {subprocess_working_directory}")
    command_string: list = list()
    try:
        for config_command in helmizer_config["helmizer"]["commandSequence"]:
            # construct command(s)
            if config_command["command"]:
                command: str = config_command["command"].get(str)
                if config_command["args"]:
                    args: str = " ".join(config_command["args"].get(list))  # combine list elements into space-delimited
                    command_string.append(f"{command} {args}")
                else:
                    command_string.append(f"{command}")

        # execute
        for command in command_string:
            if arguments.quiet:
                logging.debug(f"creating subprocess: '{command}'")
                subprocess.run(
                    f"{command}",
                    shell=True,
                    check=True,
                    stdout=subprocess.DEVNULL,
                    stderr=subprocess.DEVNULL,
                    text=True,
                    cwd=subprocess_working_directory,
                )
            else:
                logging.info(f"creating subprocess: '{command}'")
                subprocess.run(
                    f"{command}",
                    shell=True,
                    check=True,
                    text=True,
                    cwd=subprocess_working_directory,
                )
    except NotFoundError:
        pass


def init_arg_parser():
    try:
        parser = argparse.ArgumentParser(
            prog="helmizer",
            description="Helmizer",
            formatter_class=argparse.ArgumentDefaultsHelpFormatter,
        )
        args = parser.add_argument_group()
        args.add_argument(
            "--debug",
            dest="debug",
            action="store_true",
            help="enable debug logging",
            default=False,
        )
        args.add_argument(
            "--dry-run",
            dest="dry_run",
            action="store_true",
            help="do not write to a file system",
            default=False,
        )
        args.add_argument(
            "--skip-commands",
            dest="skip_commands",
            action="store_true",
            help="skip executing commandSequence, just generate kustomization file",
            default=False,
        )
        args.add_argument(
            "--no-sort-keys",
            dest="no_sort_keys",
            action="store_true",
            help="disables alphabetical sorting of sub-keys (such as 'resources') in output kustomization file",
            default=False,
        )
        args.add_argument(
            "--quiet",
            "-q",
            dest="quiet",
            action="store_true",
            help="quiet output from subprocesses",
            default=False,
        )
        args.add_argument("--version", "-v", action="version", version="v0.13.0")
        args.add_argument(
            "helmizer_config",
            action="store",
            type=str,
            help="path to helmizer config file",
        )
        arguments = parser.parse_args()

        if arguments.quiet:
            logging.basicConfig(
                level=logging.INFO,
                datefmt=None,
                stream=None,
                format="[%(asctime)s %(levelname)s] %(message)s",
            )
        if arguments.debug:
            logging.basicConfig(
                level=logging.DEBUG,
                datefmt=None,
                stream=stdout,
                format="[%(asctime)s %(levelname)s] %(message)s",
            )
        else:
            logging.basicConfig(
                level=logging.INFO,
                datefmt=None,
                stream=stdout,
                format="[%(asctime)s %(levelname)s] %(message)s",
            )

        return arguments
    except argparse.ArgumentError as e:
        logging.error("Error parsing arguments")
        raise e


def validate_helmizer_config_version(helmizer_config_version: str) -> bool:
    # TODO actually validate it
    logging.debug(f"validating helmizer config version: {helmizer_config_version}")
    return True


def init_helmizer_config(arguments) -> object:
    helmizer_config_path = arguments.helmizer_config

    config = confuse.Configuration("helmizer", __name__)
    try:
        if helmizer_config_path:
            logging.debug(f"Trying helmizer config path from argument: {helmizer_config_path}")
            config.set_file(path.normpath(helmizer_config_path))
            logging.debug(f"parsed config: {config}")

    # no config file found. Give up
    except confuse.exceptions.ConfigReadError:
        logging.error(f"Unable to locate helmizer config. Path provided: {helmizer_config_path}")
        exit(1)

    try:
        validate_helmizer_config_version(config["helmizer"]["version"].get(str))
    except NotFoundError:
        logging.debug("Unable to validate version")

    return config


def main():
    arguments = init_arg_parser()  # parse args
    helmizer_config = init_helmizer_config(arguments)  # parse config file
    if not arguments.skip_commands:  # run subprocesses
        run_subprocess(helmizer_config, arguments)
    kustomization = Kustomization(helmizer_config, arguments)  # render kustomization
    kustomization.render_template(arguments)  # instantiate kustomization YAML


if __name__ == "__main__":
    try:
        main()
    except SystemExit:
        exit(1)
    except KeyboardInterrupt:
        exit(1)

File: src/helmizer.go
package main

import (
	"bytes"
	"crypto/md5"
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	log "github.com/sirupsen/logrus"
	yaml "gopkg.in/yaml.v3"
)

// Meta struct for the Helmizer config file
type Config struct {
	Helmizer  Helmizer  `yaml:"helmizer"`
	Kustomize Kustomize `yaml:"kustomize"`
}

// Contains the subset of the Helmizer config file relating to be kustomized
type Kustomize struct {
	ApiVersion string `yaml:"apiVersion"`
	Kind       string `yaml:"kind"`
	Namespace  string `yaml:"namespace"`

	BuildMetadata         []string               `yaml:"buildMetadata"`
	CommonAnnotations     map[string]string      `yaml:"commonAnnotations"`
	CommonLabels          map[string]string      `yaml:"commonLabels"`
	ConfigMapGenerator    []interface{}          `yaml:"configMapGenerator"`
	Crds                  []string               `yaml:"crds"`
	GeneratorOptions      map[string]interface{} `yaml:"generatorOptions"`
	Images                []interface{}          `yaml:"images"`
	NamePrefix            string                 `yaml:"namePrefix"`
	NameSuffix            string                 `yaml:"nameSuffix"`
	OpenAPI               map[string]interface{} `yaml:"openapi"`
	Patches               []interface{}          `yaml:"patches"`
	PatchesJson6902       []interface{}          `yaml:"patchesJson6902"`
	PatchesStrategicMerge []string               `yaml:"patchesStrategicMerge"`
	Replacements          []interface{}          `yaml:"replacements"`
	Replicas              []interface{}          `yaml:"replicas"`
	Resources             []string               `yaml:"resources"`
	SecretGenerator       []interface{}          `yaml:"secretGenerator"`
	SortOptions           map[string]interface{} `yaml:"sortOptions"`
	Vars                  []interface{}          `yaml:"vars"`
}

// Contains the final data structure that will be marshalled to YAML for the kustomization.yaml file
type Kustomization struct {
	ApiVersion string `yaml:"apiVersion"`
	Kind       string `yaml:"kind"`
	Namespace  string `yaml:"namespace,omitempty"`

	BuildMetadata         []string               `yaml:"buildMetadata,omitempty"`
	CommonAnnotations     map[string]string      `yaml:"commonAnnotations,omitempty"`
	CommonLabels          map[string]string      `yaml:"commonLabels,omitempty"`
	ConfigMapGenerator    []interface{}          `yaml:"configMapGenerator,omitempty"`
	Crds                  []string               `yaml:"crds,omitempty"`
	GeneratorOptions      map[string]interface{} `yaml:"generatorOptions,omitempty"`
	Images                []interface{}          `yaml:"images,omitempty"`
	NamePrefix            string                 `yaml:"namePrefix,omitempty"`
	NameSuffix            string                 `yaml:"nameSuffix,omitempty"`
	OpenAPI               map[string]interface{} `yaml:"openapi,omitempty"`
	Patches               []interface{}          `yaml:"patches,omitempty"`
	PatchesJson6902       []interface{}          `yaml:"patchesJson6902,omitempty"`
	PatchesStrategicMerge []string               `yaml:"patchesStrategicMerge,omitempty"`
	Replacements          []interface{}          `yaml:"replacements,omitempty"`
	Replicas              []interface{}          `yaml:"replicas,omitempty"`
	Resources             []string               `yaml:"resources,omitempty"`
	SecretGenerator       []interface{}          `yaml:"secretGenerator,omitempty"`
	SortOptions           map[string]interface{} `yaml:"sortOptions,omitempty"`
	Vars                  []interface{}          `yaml:"vars,omitempty"`
}

// Contains the subset of the Helmizer config file relating to either command sequence
type Commands struct {
	Arguments []string `yaml:"args"`
	Command   string   `yaml:"command"`
}

// Contains the subset of the Helmizer config file
type Helmizer struct {
	// CLI args only
	LogColors bool   `yaml:"logColors"`
	LogFormat string `yaml:"logFormat"`
	LogLevel  string `yaml:"logLevel"`

	// These don't exist with CLI args
	PostCommands []Commands `yaml:"postCommands"`
	PreCommands  []Commands `yaml:"preCommands"`

	// CLI args override config equivalents
	ApiVersion        string   `yaml:"apiVersion"`
	DryRun            bool     `yaml:"dryRun"`
	Ignore            []string `yaml:"ignore"`
	KustomizationPath string   `yaml:"KustomizationPath"`
	QuietCommands     bool     `yaml:"quietCommands"`
	SkipAllCommands   bool     `yaml:"skipAllCommands"`
	SkipPostCommands  bool     `yaml:"skipPostCommands"`
	SkipPreCommands   bool     `yaml:"skipPreCommands"`
	StopOnError       bool     `yaml:"stopOnError"`
}

// ExecuteCommands executes the command sequence specified in the Helmizer configuration.
func ExecuteCommands(stage string, args CLIArgs, helmizer Helmizer) ([]string, error) {
	var outputs []string

	// Determine which command sequence to execute
	commands := []Commands{}
	if stage == "pre" {
		commands = helmizer.PreCommands
	} else if stage == "post" {
		commands = helmizer.PostCommands
	}

	for _, command := range commands {
		cmd := exec.Command(command.Command, command.Arguments...)

		// Combine the command with its arguments
		commandString := strings.Join(cmd.Args, " ")

		// Set the working directory
		var err error
		cmd.Dir, err = filepath.Abs(filepath.Dir(args.ConfigFilePath))
		if err != nil {
			log.Errorf("Failed to set working directory: %s", err)
			return outputs, err
		}

		// Prepare to capture both stdout and stderr
		var stdoutStderr bytes.Buffer

		// Evaluate if command output should be streamed out
		if !helmizer.QuietCommands {
			cmd.Stdout = &stdoutStderr
			cmd.Stderr = &stdoutStderr
		}

		// Run the command
		err = cmd.Run()
		if err != nil {
			log.Errorf("Failed to execute command: %s\n%s", commandString, stdoutStderr.String())
			if helmizer.StopOnError {
				log.Fatal("stopOnError is set to true - exiting")
				os.Exit(1)
			} else {
				log.Debug("Command failed, but continuing on")
				return outputs, err
			}
		} else {
			// Log the output, trimming command
			output := strings.TrimSuffix(stdoutStderr.String(), "\n")
			log.Infof("Successfully executed command: %s\n%s", commandString, output)
			outputs = append(outputs, output)
		}
	}
	return outputs, nil
}

// Retrieves a list of files or URLs based on the provided configuration file path and configuration.
// It returns a slice of strings containing the absolute or relative paths to the files.
func GetFilesOrURLs(Type string, ConfigFilePath string, config Config) []string {
	var outputList []string // Contains list of URLs or files
	// Get absolute path to parent directory containing helmizer.yaml
	HelmizerConfigAbsDir, _ := filepath.Abs(filepath.Dir(ConfigFilePath))

	typeToIterateOver := []string{}

	if Type == "Crds" {
		typeToIterateOver = config.Kustomize.Crds
	} else if Type == "PatchesStrategicMerge" {
		typeToIterateOver = config.Kustomize.PatchesStrategicMerge
	} else if Type == "Resources" {
		typeToIterateOver = config.Kustomize.Resources
	}

	for _, KustomizeFilePath := range typeToIterateOver {
		var absPathToWalk string
		var isAbsolutePath bool

		// Check if the path is a URL
		if strings.HasPrefix(KustomizeFilePath, "http://") || strings.HasPrefix(KustomizeFilePath, "https://") {
			log.Debugf("Skipping URL: %s", KustomizeFilePath)
			outputList = append(outputList, KustomizeFilePath)
			continue // Skip this URL
		}

		// Check if the path is absolute
		if filepath.IsAbs(KustomizeFilePath) {
			absPathToWalk = KustomizeFilePath
			isAbsolutePath = true
		} else {
			// Convert relative path to absolute path based on HelmizerConfigAbsDir
			var err error
			absPathToWalk, err = filepath.Abs(filepath.Join(HelmizerConfigAbsDir, KustomizeFilePath))
			if err != nil {
				log.Errorf("Error converting to absolute path: %s", err)
				continue // Skip this path
			}
			isAbsolutePath = false
		}

		// Check if the directory exists
		if _, err := os.Stat(absPathToWalk); os.IsNotExist(err) {
			log.Debugf("Directory does not exist: %s", absPathToWalk)
			continue // Skip this directory
		}

		// Walk the directory, accumulating a list of files
		err := filepath.Walk(absPathToWalk, func(pathToFile string, info os.FileInfo, err error) error {
			if err != nil {
				log.Errorf("Error accessing path %q: %v", pathToFile, err)
				return err
			}

			if !info.IsDir() {
				log.Debugf("Found file: %s", pathToFile)
				if isAbsolutePath {
					outputList = append(outputList, pathToFile)
				} else {
					// Convert the absolute path to a relative path
					relativePath, err := filepath.Rel(HelmizerConfigAbsDir, pathToFile)
					if err != nil {
						log.Errorf("Error converting to relative path: %s", err)
						return err
					}
					outputList = append(outputList, relativePath)
				}
			}
			return nil
		})

		if err != nil {
			log.Warnf("Error walking the path %q: %v", absPathToWalk, err)
		}
	}
	outputList = CompareAndStrip(config.Helmizer.Ignore, outputList)
	return outputList
}

// RenameHelmizerKeys reads the Helmizer configuration from a YAML file and renames the keys
func RenameHelmizerKeys(filePath string) error {
	// Read the file
	data, err := os.ReadFile(filePath)
	if err != nil {
		log.Error(err)
		return err
	}

	// Unmarshal the YAML data into a map
	var config map[string]interface{}
	err = yaml.Unmarshal(data, &config)
	if err != nil {
		log.Error(err)
		return err
	}

	// Get the Helmizer keys
	helmizer, ok := config["helmizer"].(map[string]interface{})
	if !ok {
		log.Debug("No 'helmizer' key found in the YAML file")
	}

	// Loop through the Helmizer keys and print logs based on the key names
	for key, value := range helmizer {
		switch key {
		case "dry-run":
			delete(helmizer, "dry-run")
			helmizer["DryRun"] = value
			log.Warn("Pre-flight: Helmizer config 'dry-run' was replaced with 'dryRun'. Update your configuration. This will be removed in a future release")
		case "commandSequence":
			delete(helmizer, "commandSequence")
			helmizer["preCommands"] = value
			log.Warn("Pre-flight: Helmizer config 'commandSequence' was replaced with 'preCommands'. Update your configuration. This will be removed in a future release")
		case "kustomization-file-extension", "kustomization-file-name", "sort-keys", "version":
			log.Warnf("Pre-flight: Helmizer config '%s' is deprecated and will be removed in a future release Update your configuration.", key)
		}
	}

	return nil
}

// Reconciles the Helmizer configuration with the CLI arguments and returns a joined config
func ReconcileHelmizerConfig(args CLIArgs, config Config) Config {
	_ = RenameHelmizerKeys(args.ConfigFilePath)

	if config.Helmizer.ApiVersion == "" {
		config.Helmizer.ApiVersion = args.ApiVersion
	}

	if args.DryRun {
		config.Helmizer.DryRun = true
	}

	if config.Helmizer.KustomizationPath == "" {
		config.Helmizer.KustomizationPath = args.KustomizationPath
	}

	if args.LogFormat != "" {
		config.Helmizer.LogFormat = args.LogFormat
	}

	if args.LogLevel != "" {
		config.Helmizer.LogLevel = args.LogLevel
	}

	if args.QuietCommands {
		config.Helmizer.QuietCommands = true
	}

	if !args.StopOnError {
		config.Helmizer.StopOnError = false
	}

	if args.SkipAllCommands {
		config.Helmizer.SkipAllCommands = true
	}

	if args.SkipPostCommands {
		config.Helmizer.SkipPostCommands = true
	}

	if args.SkipPostCommands {
		config.Helmizer.SkipPostCommands = true
	}

	if args.SkipPreCommands {
		config.Helmizer.SkipPreCommands = true
	}

	return config
}

func WriteKustomizationFile(args CLIArgs, helmizer Config, kustomization Kustomization, kFilePath string) {
	// Get the absolute path to the kustomization file
	kAbsPath, _ := filepath.Abs(kFilePath)

	// Read the existing file
	oldContent, err := os.ReadFile(kAbsPath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			// If the file doesn't exist, set oldContent to an empty byte slice
			log.Debugf("File %s does not exist, proceeding with empty content", kAbsPath)
			oldContent = []byte{}
		} else {
			// If some other error occurred, log it and exit
			log.Fatal(err)
		}
	}

	// Fix indentation of the old content
	fixedOldContent := FixYAMLIndentation(oldContent)

	// Calculate MD5 hash of the fixed old content
	oldHash := md5.Sum(fixedOldContent)

	// Marshal kustomization to YAML
	kustomizationYAML, err := yaml.Marshal(&kustomization)
	if err != nil {
		log.Fatal(err)
	}

	// Fix indentation of the new content
	fixedNewContent := FixYAMLIndentation(kustomizationYAML)

	// Calculate MD5 hash of the fixed new content
	newHash := md5.Sum(fixedNewContent)

	// Compare old hash with new hash
	if oldHash != newHash {
		// Write new content to file
		err = os.WriteFile(kAbsPath, fixedNewContent, 0644)
		if err != nil {
			log.Fatal(err)
		} else {
			log.Printf("Successfully wrote kustomization file: %s", kAbsPath)
		}
	} else {
		log.Infof("kustomization file's keys/values have not changed - not writing file: %s", kAbsPath)
	}
	log.Infof("Helmizer completed successfully")
}
